src/main/java/org/broadinstitute/hellbender/tools/HaplotypeCallerSpark.java

```diff
@@ -0,0 +1,329 @@
+        try(final VariantContextWriter writer = hcEngine.makeVCFWriter(output, this.getBestAvailableSequenceDictionary())) {
+            hcEngine.writeHeader(writer);
+            sortedVariants.forEach(writer::add);
+        }
+    }
+
+    /**
+     * Create an RDD of {@link ReadShard} from an RDD of {@link GATKRead}
+     * @param shardBoundariesBroadcast  broadcast of an {@link OverlapDetector} loaded with the intervals that should be used for creating ReadShards
+     * @param reads Rdd of {@link GATKRead}
+     * @return a Rdd of reads grouped into potentially overlapping shards
+     */
+    private static JavaRDD<ReadShard> createReadShards(final Broadcast<OverlapDetector<ReadShard.ShardBoundary>> shardBoundariesBroadcast, final JavaRDD<GATKRead> reads) {
+        final JavaPairRDD<ReadShard.ShardBoundary, GATKRead> paired = reads.flatMapToPair(read -> {
+            final Collection<ReadShard.ShardBoundary> overlappingShards = shardBoundariesBroadcast.value().getOverlaps(read);
+            return overlappingShards.stream().map(key -> new Tuple2<>(key, read)).collect(Collectors.toList());
+        });
+        final JavaPairRDD<ReadShard.ShardBoundary, Iterable<GATKRead>> shardsWithReads = paired.groupByKey();
+        return shardsWithReads.map(shard -> new SparkReadShard(shard._1(), shard._2()));
+    }
+
+    /**
+     * @return an {@link OverlapDetector} loaded with {@link org.broadinstitute.hellbender.engine.ReadShard.ShardBoundary}
+     * based on the -L intervals
+     */
+    private static OverlapDetector<ReadShard.ShardBoundary> getShardBoundaryOverlapDetector(final SAMFileHeader header, final List<SimpleInterval> intervals, final int readShardSize, final int readShardPadding) {
+        final List<ReadShard.ShardBoundary> keys = intervals.stream()
+                .flatMap(interval -> ReadShard.divideIntervalIntoShards(interval, readShardSize, readShardSize, readShardPadding, header.getSequenceDictionary()).stream())
+                .collect(Collectors.toList());
+        return OverlapDetector.create(keys);
+    }
+
+    /**
+     * @return and RDD of {@link Tuple2<AssemblyRegion,SimpleInterval>} which pairs each AssemblyRegion with the
+     * interval it was generated in
+     */
+    private static FlatMapFunction<ReadShard, Tuple2<AssemblyRegion, SimpleInterval>> shardsToAssemblyRegions(
+            final AuthHolder authHolder,
+            final Broadcast<ReferenceMultiSource> reference,
+            final Broadcast<HaplotypeCallerArgumentCollection> hcArgsBroadcast,
+            final ShardingArgumentCollection assemblyArgs,
+            final SAMFileHeader header) {
+        return shard -> {
+            final ReferenceMultiSource referenceMultiSource = reference.value();
+            final ReferenceMultiSourceToReferenceDataSourceAdaptor refDataSource = new ReferenceMultiSourceToReferenceDataSourceAdaptor(referenceMultiSource, authHolder);
```

Comment:
This is a longstanding problem, and one that needs to be addressed.  I think it's out of scope of this pull request to fix the proliferation of reference data types though.

I did unify the adaptors classes though, since they're so similar.  (which implies that we really should be able to unify these interfaces...)